### ISSUE TITLE
Training from scratch does not work

### ISSUE THREAD

**Issue by mrastegari** [facebook/fb.resnet.torch#17]
I tried to train the resnet-18 from scratch but I could not get the same accuracy. my top-1 is = (val 58.6%, train 54.7%) . I am using Titan-X  I am wondering if that matters? I also noticed that iteration time varries a lot and as it goes more iterations it become slower. Mostly it happen in Dataloading time.

**Comment by colesbury** [comment]
1. Make sure you are running the with the latest cunn/nn/cudnn. What flags are you using to train?
2. Increase the number of data loading threads until the data loading time is consistently small. Are your images on an SSD?

**Comment by mrastegari** [comment]
All the libraries are updated .Yes everything is on SSD.  I put the outputs for 8 threads and 16 threads. You can see in both cases after a while it will get very unstable (in terms of time).  

th main.lua -depth 18 -nGPU 4 -nThreads 8 -batchSize 256 -data ~/data/imagenet2012/
=> Creating model from file: models/resnet.lua	
 | ResNet-18 ImageNet	
=> Training epoch # 1	
 | Epoch: [1][1/5005]    Time 21.933  Data 10.801  Err 7.0894  top1 100.000  top5 100.000	
 | Epoch: [1][2/5005]    Time 0.426  Data 0.002  Err 7.0478  top1 100.000  top5  99.609	
 | Epoch: [1][3/5005]    Time 1.147  Data 0.000  Err 7.0691  top1  99.609  top5  99.609	
 | Epoch: [1][4/5005]    Time 0.543  Data 0.001  Err 7.0632  top1 100.000  top5  99.219	
 | Epoch: [1][5/5005]    Time 0.375  Data 0.001  Err 7.0304  top1 100.000  top5 100.000	
 | Epoch: [1][6/5005]    Time 1.369  Data 0.001  Err 7.0570  top1  99.609  top5  99.219	
 | Epoch: [1][7/5005]    Time 0.453  Data 0.001  Err 7.1039  top1 100.000  top5 100.000	
 | Epoch: [1][8/5005]    Time 1.204  Data 0.001  Err 7.0371  top1 100.000  top5  99.219	
 | Epoch: [1][9/5005]    Time 0.520  Data 0.024  Err 7.1100  top1 100.000  top5 100.000	
 | Epoch: [1][10/5005]    Time 0.347  Data 0.000  Err 7.1325  top1  99.609  top5  98.828	
 | Epoch: [1][11/5005]    Time 0.451  Data 0.037  Err 7.1181  top1 100.000  top5 100.000	
 | Epoch: [1][12/5005]    Time 0.415  Data 0.000  Err 7.0857  top1 100.000  top5  99.609	
 | Epoch: [1][13/5005]    Time 0.242  Data 0.000  Err 7.0709  top1 100.000  top5  98.047	
 | Epoch: [1][14/5005]    Time 0.299  Data 0.006  Err 7.1011  top1 100.000  top5  99.609	
 | Epoch: [1][15/5005]    Time 0.665  Data 0.008  Err 7.1810  top1 100.000  top5  99.219	
 | Epoch: [1][16/5005]    Time 0.250  Data 0.000  Err 7.1756  top1  99.609  top5  98.828	
 | Epoch: [1][17/5005]    Time 4.851  Data 4.148  Err 7.1580  top1 100.000  top5 100.000	
 | Epoch: [1][18/5005]    Time 2.142  Data 0.001  Err 7.1383  top1 100.000  top5 100.000	
 | Epoch: [1][19/5005]    Time 2.286  Data 0.001  Err 7.0600  top1  99.609  top5  99.219	
 | Epoch: [1][20/5005]    Time 0.612  Data 0.000  Err 7.0960  top1  99.609  top5  98.438	
 | Epoch: [1][21/5005]    Time 1.365  Data 0.911  Err 7.0929  top1 100.000  top5 100.000	
 | Epoch: [1][22/5005]    Time 2.598  Data 1.229  Err 7.0373  top1 100.000  top5  99.609	
 | Epoch: [1][23/5005]    Time 0.734  Data 0.004  Err 7.1312  top1 100.000  top5  99.219	
 | Epoch: [1][24/5005]    Time 3.368  Data 2.928  Err 7.1233  top1  99.609  top5  99.609	
 | Epoch: [1][25/5005]    Time 5.038  Data 4.212  Err 7.1015  top1 100.000  top5  99.609	
 | Epoch: [1][26/5005]    Time 0.374  Data 0.059  Err 7.1453  top1 100.000  top5  99.219	
 | Epoch: [1][27/5005]    Time 1.350  Data 0.187  Err 7.0581  top1  99.609  top5  98.828	
 | Epoch: [1][28/5005]    Time 0.424  Data 0.001  Err 7.0892  top1 100.000  top5  99.609	
 | Epoch: [1][29/5005]    Time 2.760  Data 2.113  Err 7.0377  top1 100.000  top5  99.609	
 | Epoch: [1][30/5005]    Time 0.621  Data 0.158  Err 7.0433  top1  99.609  top5  99.219	
 | Epoch: [1][31/5005]    Time 0.516  Data 0.134  Err 7.0316  top1  99.219  top5  98.438	
 | Epoch: [1][32/5005]    Time 2.786  Data 2.420  Err 7.0360  top1 100.000  top5 100.000	
 | Epoch: [1][33/5005]    Time 3.392  Data 2.986  Err 6.9629  top1  99.219  top5  99.219	
 | Epoch: [1][34/5005]    Time 1.852  Data 1.142  Err 6.9965  top1  99.609  top5  99.609	
 | Epoch: [1][35/5005]    Time 1.312  Data 0.955  Err 6.9420  top1  99.219  top5  99.219	
 | Epoch: [1][36/5005]    Time 1.124  Data 0.001  Err 7.0049  top1 100.000  top5  99.219	
 | Epoch: [1][37/5005]    Time 5.034  Data 2.718  Err 6.9277  top1  99.219  top5  98.438	
 | Epoch: [1][38/5005]    Time 0.397  Data 0.001  Err 6.9698  top1 100.000  top5  99.609	
 | Epoch: [1][39/5005]    Time 0.844  Data 0.070  Err 6.9385  top1  99.609  top5  98.438	
 | Epoch: [1][40/5005]    Time 3.053  Data 2.655  Err 7.0031  top1 100.000  top5  99.609	


HERE is with 16 threads:

th main.lua -depth 18 -nGPU 4 -nThreads 16 -batchSize 256 -data ~/data/imagenet2012/
=> Creating model from file: models/resnet.lua	
 | ResNet-18 ImageNet	
=> Training epoch # 1	
 | Epoch: [1][1/5005]    Time 43.409  Data 17.438  Err 7.0884  top1 100.000  top5 100.000	
 | Epoch: [1][2/5005]    Time 0.894  Data 0.003  Err 7.0561  top1 100.000  top5 100.000	
 | Epoch: [1][3/5005]    Time 1.513  Data 0.000  Err 7.0450  top1  99.609  top5  99.609	
 | Epoch: [1][4/5005]    Time 0.760  Data 0.000  Err 7.0696  top1 100.000  top5  99.219	
 | Epoch: [1][5/5005]    Time 1.445  Data 0.001  Err 7.0495  top1  99.609  top5  99.219	
 | Epoch: [1][6/5005]    Time 2.069  Data 0.005  Err 7.0352  top1  99.609  top5  99.219	
 | Epoch: [1][7/5005]    Time 1.214  Data 0.002  Err 6.9744  top1  99.609  top5  98.438	
 | Epoch: [1][8/5005]    Time 1.182  Data 0.001  Err 7.1251  top1 100.000  top5  98.828	
 | Epoch: [1][9/5005]    Time 2.298  Data 0.000  Err 7.0085  top1  99.609  top5  98.828	
 | Epoch: [1][10/5005]    Time 1.497  Data 0.016  Err 7.1110  top1 100.000  top5 100.000	
 | Epoch: [1][11/5005]    Time 1.207  Data 0.001  Err 7.1388  top1  99.609  top5  99.219	
 | Epoch: [1][12/5005]    Time 1.118  Data 0.001  Err 7.2170  top1 100.000  top5  99.609	
 | Epoch: [1][13/5005]    Time 2.231  Data 0.001  Err 7.0511  top1 100.000  top5  98.828	
 | Epoch: [1][14/5005]    Time 1.579  Data 0.004  Err 7.1527  top1  99.219  top5  99.219	
 | Epoch: [1][15/5005]    Time 0.942  Data 0.001  Err 7.1245  top1  99.609  top5  99.219	
 | Epoch: [1][16/5005]    Time 1.841  Data 0.001  Err 7.1061  top1  99.219  top5  98.828	
 | Epoch: [1][17/5005]    Time 1.963  Data 0.020  Err 7.0852  top1 100.000  top5  99.609	
 | Epoch: [1][18/5005]    Time 1.274  Data 0.001  Err 7.1144  top1  99.609  top5  98.828	
 | Epoch: [1][19/5005]    Time 1.659  Data 0.001  Err 7.0548  top1  99.609  top5  98.047	
 | Epoch: [1][20/5005]    Time 1.165  Data 0.000  Err 7.1616  top1 100.000  top5  99.219	
 | Epoch: [1][21/5005]    Time 0.958  Data 0.011  Err 7.0599  top1  99.609  top5  98.828	
 | Epoch: [1][22/5005]    Time 1.089  Data 0.000  Err 7.0256  top1 100.000  top5  98.828	
 | Epoch: [1][23/5005]    Time 1.384  Data 0.001  Err 7.0641  top1  99.609  top5  99.219	
 | Epoch: [1][24/5005]    Time 0.921  Data 0.000  Err 7.0724  top1 100.000  top5  99.609	
 | Epoch: [1][25/5005]    Time 1.640  Data 0.002  Err 7.0861  top1 100.000  top5  99.219	
 | Epoch: [1][26/5005]    Time 0.937  Data 0.001  Err 7.0694  top1 100.000  top5  99.219	
 | Epoch: [1][27/5005]    Time 1.081  Data 0.002  Err 7.0341  top1 100.000  top5  99.609	
 | Epoch: [1][28/5005]    Time 2.016  Data 0.004  Err 7.0488  top1  99.609  top5  99.219	
 | Epoch: [1][29/5005]    Time 1.133  Data 0.004  Err 7.0827  top1  99.609  top5  99.219	
 | Epoch: [1][30/5005]    Time 1.962  Data 0.002  Err 7.0540  top1 100.000  top5  99.219	
 | Epoch: [1][31/5005]    Time 0.963  Data 0.001  Err 7.0036  top1 100.000  top5  99.609	
 | Epoch: [1][32/5005]    Time 1.098  Data 0.001  Err 6.9691  top1 100.000  top5  99.219	
 | Epoch: [1][33/5005]    Time 4.845  Data 0.001  Err 6.9793  top1  99.609  top5  99.219	
 | Epoch: [1][34/5005]    Time 1.647  Data 0.001  Err 6.8951  top1  99.219  top5  97.266	
 | Epoch: [1][35/5005]    Time 4.597  Data 0.001  Err 6.9398  top1  99.219  top5  98.438	
 | Epoch: [1][36/5005]    Time 0.851  Data 0.001  Err 6.9511  top1  99.609  top5  99.609	
 | Epoch: [1][37/5005]    Time 1.314  Data 0.001  Err 7.0545  top1 100.000  top5  99.219	
 | Epoch: [1][38/5005]    Time 2.534  Data 0.002  Err 6.9338  top1 100.000  top5 100.000	
 | Epoch: [1][39/5005]    Time 2.206  Data 0.000  Err 6.9960  top1  99.609  top5  98.828	
 | Epoch: [1][40/5005]    Time 1.485  Data 0.000  Err 6.9310  top1  99.219  top5  98.438

**Comment by colesbury** [comment]
The data loading time with 16 threads looks OK, but the variation per-mini-batch is strange. On a Kepler GPU, the time per mini-batch is about 0.55-0.6 seconds. I would expect it to be about twice as fast on a Titan X.

What's the CPU on the machine you are using?

Here are the expected train loss and train and validation errors at the end of every epoch:

Train loss:
<img width="861" alt="resnet-18-train-loss" src="https://cloud.githubusercontent.com/assets/655866/13410081/e3cb4f28-df03-11e5-8035-330840e8459e.png">

Train error:
<img width="834" alt="resnet-18-train-err" src="https://cloud.githubusercontent.com/assets/655866/13410085/e9a90520-df03-11e5-9e2d-7fc721bdd87a.png">

Validation error:
<img width="835" alt="resnet-18-test-err" src="https://cloud.githubusercontent.com/assets/655866/13410106/fd99c790-df03-11e5-8f82-a732aa6fec0d.png">

**Comment by omkarparkhi** [comment]
Hi, I am experiencing a similar problem. My top-1 training error after 39th iteration for 18 layer model is 74%. Using 2 GPUs are 4 data loading threads.

**Comment by colesbury** [comment]
What batch size are and learning rate are you using?

**Comment by omkarparkhi** [comment]
Originally 256 for the values reported above. I also tried 128 with a similar result. I will try training on 4 GPUs later today and revert. Also, are you preprocessing the images to have minimum dimension resized to 256 pixels?

**Comment by omkarparkhi** [comment]
After resizing the training images to have smallest dimension scaled to 256 and using 4 GPU machine, I am managing to match your plot. Will revert after the training completes.

**Comment by omkarparkhi** [comment]
Training res-net 18 finished. With 4 titan -x and batch size of 256, took little over 2 days to finish.
Final numbers are as follows:

* Finished top1: 30.578  top5: 11.081

@colesbury, Thanks for the help.

**Comment by colesbury** [comment]
@omkarparkhi, no I've only used 8 GPUs for the 152-layer model

**Comment by colesbury** [comment]
@mrastegari, now that you published a paper I assume training works? What turned out to be the problem?

**Comment by pedropgusmao** [comment]
Hi, 
I am training resnet-34 on a single TitanX and I think I'm getting lower validation accuracy than expected (when compared to graphs in https://github.com/facebook/fb.resnet.torch/blob/master/pretrained/CONVERGENCE.md ).
 
I am issuing the following command:
`CUDA_VISIBLE_DEVICES=1 th main.lua -data /datasets/imagenet/cls-loc/ -dataset imagenet -gen /data/pedro/networks/resnet34/ -nThreads 8 -batchSize 128 -depth 34 > resnet34_acc.txt`

And getting the following accuracies for the first 16 epochs (which I believe should be below 50%, right? )

* Finished epoch # 1     top1:  81.274  top5:  59.467
 * Finished epoch # 2     top1:  69.453  top5:  42.166
 * Finished epoch # 3     top1:  63.817  top5:  36.616
 * Finished epoch # 4     top1:  62.510  top5:  34.728
 * Finished epoch # 5     top1:  62.754  top5:  35.115
 * Finished epoch # 6     top1:  61.130  top5:  34.197
 * Finished epoch # 7     top1:  63.295  top5:  35.920
 * Finished epoch # 8     top1:  61.608  top5:  34.487
 * Finished epoch # 9     top1:  60.412  top5:  33.195
 * Finished epoch # 10     top1:  58.558  top5:  30.735
 * Finished epoch # 11     top1:  57.717  top5:  29.971
 * Finished epoch # 12     top1:  58.314  top5:  30.744
 * Finished epoch # 13     top1:  57.246  top5:  30.006
 * Finished epoch # 14     top1:  55.973  top5:  28.523
 * Finished epoch # 15     top1:  57.394  top5:  29.909
 * Finished epoch # 16     top1:  59.061  top5:  31.899

Am I doing anything wrong?

**Comment by colesbury** [comment]
@pedropgusmao, you may need to reduce the initial learning rate (`-LR`), since the batch size is smaller. (I trained all models with batch size 256). Maybe try `-LR 0.05` or `-LR 0.075`?

**Comment by omkarparkhi** [comment]
@colesbury , My 152 training on 4 GPUs. *Finished top1: 23.495  top5:  6.803.*  
I didn't find your numbers anywhere.

**Comment by pedropgusmao** [comment]
@colesbury ,  Thanks for the tip. I continued training the network using `-LR 0.1` just to see what happened.  I am now at epoch 53 and things look better. Validation error is now around 36% =) . Again, thanks.

**Comment by colesbury** [comment]
@omkarparkhi: ResNet-152 on 8 GPUs: `Finished top1: 22.264  top5:  6.237` What batch size and LR schedule are you using?

**Comment by colesbury** [comment]
@omkarparkhi, I've added a [pretrained ResNet-152](https://github.com/facebook/fb.resnet.torch/tree/master/pretrained) model in-case that's helpful. The validation error (22.16%/6.16%) is slightly better than I wrote above, because I recomputed the batch norm statistics on a random subset of the training data after the model finished training.

**Comment by pedropgusmao** [comment]
Does anyone know why validation error is lower than training error? I am getting the same results, but it's an odd behaviour, right ?

**Comment by colesbury** [comment]
It's because of the scale and aspect ratio jittering. The training is on distorted images, while the validation is on the center crops.

**Comment by ljk628** [comment]
Had exactly the same issue with Intel Core i7-5930K (6 cores). Just want to add that this issue was solved by setting OMP_NUM_THREADS=1 as suggested in https://github.com/facebook/fb.resnet.torch/issues/12 . Otherwise the CPU is not fully utilized and significantly slows down the loading speed. I would suggest to add this "trick" in README.

**Comment by colesbury** [comment]
That's great! I think we can set that automatically via torch.setnumthreads in the code. I'll check that it works and add it later this week

**Comment by ljk628** [comment]
Yup. Just find an old issue might be related https://github.com/torch/image/issues/7 .  If simply comment the preprocessing part https://github.com/facebook/fb.resnet.torch/blob/a446597550f7e4b15866e07b32c21e63cd45064b/datasets/imagenet.lua#L84-L90 , it will run smoothly even without the OMP flag.
